pkg/executor/sortexec/sort.go

```diff
@@ -119,57 +122,173 @@ func (e *SortExec) Next(ctx context.Context, req *chunk.Chunk) error {
 		e.fetched = true
 	}
 
-	if len(e.PartitionList) == 0 {
+	sortPartitionListLen := len(e.sortPartitions)
+	if sortPartitionListLen == 0 {
 		return nil
 	}
-	if len(e.PartitionList) > 1 {
-		if err := e.externalSorting(req); err != nil {
+
+	if sortPartitionListLen == 1 {
+		if err := e.nonExternalSorting(req); err != nil {
 			return err
 		}
 	} else {
-		for !req.IsFull() && e.Idx < e.PartitionList[0].NumRow() {
-			_, _, err := e.PartitionList[0].GetSortedRowAndAlwaysAppendToChunk(e.Idx, req)
-			if err != nil {
-				return err
+		if err := e.externalSorting(req); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *SortExec) initExternalSorting() error {
+	err := e.initCursors()
+	if err != nil {
+		return err
+	}
+
+	e.multiWayMerge = &multiWayMerge{e.lessRow, e.compressRow, make([]rowWithPartition, 0, len(e.cursors))}
+	for i := 0; i < len(e.cursors); i++ {
+		// We should always get row here
+		row := e.cursors[i].getSpilledRow()
+		e.cursors[i].advanceRow()
+		e.multiWayMerge.elements = append(e.multiWayMerge.elements, rowWithPartition{row: *row, partitionID: i})
+	}
+	heap.Init(e.multiWayMerge)
+	return nil
+}
+
+func (e *SortExec) nonExternalSorting(req *chunk.Chunk) (err error) {
+	e.helper.syncLock.Lock()
+	defer e.helper.syncLock.Unlock()
+
+	if e.helper.spillError != nil {
+		return e.helper.spillError
+	}
+
+	if e.sortPartitions[0].isSpillTriggered() {
+		if len(e.cursors) == 0 {
+			e.initCursors()
+		}
+
+		// Maybe the spill is triggered by the last chunk, so there is only one partition.
```

Comment:
`nonExternalSorting` handling spilled partitions can be confusing.

Reply:
> `nonExternalSorting` handling spilled partitions can be confusing.

I will change its name to `onePartitionSorting`.